src/Perlang.Stdlib/Lang/AsciiString.cs

```diff
@@ -0,0 +1,228 @@
+    /// array will be reused; no copying will take place.
+    /// </summary>
+    /// <param name="bytes">A pointer to a byte array, validated by the caller to be ASCII safe.</param>
+    /// <param name="length">The length of the string, excluding NUL terminator.</param>
+    /// <returns>A newly allocated <see cref="AsciiString"/>.</returns>
+    private static unsafe AsciiString from(byte* bytes, nuint length)
+    {
+        return new AsciiString(bytes, length);
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AsciiString"/> class from an existing <see cref="char"/> array. A
+    /// new `byte[]` will be allocated to contain the string content.
+    /// </summary>
+    /// <param name="chars">A char array, validated by the caller to be ASCII safe.</param>
+    /// <returns>A newly allocated <see cref="AsciiString"/>.</returns>
+    public static AsciiString from(IReadOnlyList<char> chars)
+    {
+        return new AsciiString(chars);
+    }
+
+    /// <summary>
+    /// Creates a new <see cref="AsciiString"/> from the given .NET string. Note that this method will allocate a new
+    /// backing `byte[]` for holding the string contents; it will not reuse the existing bytes.
+    /// </summary>
+    /// <param name="s">The input string.</param>
+    /// <returns>A newly allocated <see cref="AsciiString"/>.</returns>
+    /// <exception cref="ArgumentException">The given string contains non-ASCII characters.</exception>
+    public static AsciiString from(string s)
+    {
+        char[] chars = s.ToCharArray();
+
+        foreach (char c in chars)
+        {
+            if (c > 127)
+            {
+                throw new ArgumentException($"Attempted to construct an AsciiString from a string which contains non-ASCII characters: {s}");
+            }
+        }
+
+        return new AsciiString(chars);
+    }
+
+    private unsafe AsciiString(byte* bytes, nuint length)
+    {
+        this.bytes = bytes;
+        this.length = length;
+
+        // We presume the caller to provide us a valid string, but since the absence of this can cause so nefarious
+        // bugs, we validate it when running in debug mode.
+        Debug.Assert(bytes[length] == 0, $"String is not properly NUL-terminated");
+    }
+
+    // TODO: Try to get rid of this constructor, once we have rewritten the scanner to use Perlang strings instead
+    // TODO: of .NET strings.
+
+    private unsafe AsciiString(IReadOnlyList<char> chars)
+    {
+        // TODO: Use jemalloc instead. Requires creating a NuGet package of it for easiest consumption from this project.
+        // TODO: See https://github.com/dotnet/runtime/issues/11404 and https://github.com/allisterb/jemalloc.NET for
+        // TODO: some prior art.
+        //
+        // TODO: Use a wrapper for keeping track of our allocations, and free all unmanaged memory on VM exit. Something
+        // TODO: like AppDomain.CurrentDomain.DomainUnload could probably do.
```

Comment:
TODO: This one needs to be taken care of _before_ this PR is merged.

Reply:
Fixed & tested locally. We now have some basic logic in place which "garbage collects" at the end of the process. This is clearly unsuitable for many "real-world" applications, but we'll defer fixing this properly until #378 is taken care of. Considering "done" for now.